feat(codegen): fix module discovery logic

bjaanes opened the pull request:
Issue: #835 

# Background
The code generation for vuex was not able to find all modules due to the nature of the algorithm used to discover modules. Previously it checked for an implementation of the msg interface which would signal that the code was indeed a cosmos sdk-based module.

During the development I found no simple way of determining for a non-msg module (query only) such as [cosmos/tx/v1beta1](https://github.com/cosmos/cosmos-sdk/tree/master/proto/cosmos/tx/v1beta1) if it was a proper sdk-based module or just any other proto files. So the approach has changed a little, but that also means it covers another issue which was mentioned in the code about generating code for non-registered modules (those are now skipped)

# What has been done
- Added a function in pkg/cosmosanlysis/app to find all registered modules (see comments for details on the algorithm)
- Changed the aforementioned algorithm in pkg/cosmosanlysis/module to use the registered modules as both a filter and a way to look up the implemented code to check if the proto files are implemented there
- Written both unit and integration tests to cover the changes + expand on the module test with more useful test data

Bounty address in commit message.


Note: I would like some extra scrutiny on my ast code (pkg/cosmosanlysis/app/app.go) since it is the first time I am doing ast parsing in go. I found it a bit more verbose and "indenty" than I would've liked. Perhaps someone with more experience can give me pointers to make it better!

Lastly, thanks for the bounty opportunity. I hope this helps and I am happy to jump on more the future :)

ilgooz commented on starport/pkg/cosmosanalysis/module/module.go:
```diff
@@ -90,17 +91,39 @@ func Discover(ctx context.Context, sourcePath, protoDir string) ([]Module, error
 		return nil, err
 	}
 
+	rm, err := app.GetRegisteredModules(filepath.Join(chainRoot, "app"))
```
Btw, can we move `app` into a constant in the beginning of this file?

ilgooz reviewed (changes requested):
Hey, I made a quick review. Looks great! 🎉

bjaanes commented on integration/cosmosgen/cosmosgen_test.go:
```diff
@@ -0,0 +1,146 @@
+	}
+
+	for _, chainModule := range expectedChainModules {
+		_, statErr := os.Stat(filepath.Join(path, "vue", "src", "store", "generated", "test", "blog", chainModule))
```
That makes a lot sense, doing that too.

bjaanes commented on integration/cosmosgen/cosmosgen_test.go:
```diff
@@ -0,0 +1,146 @@
+
+	for _, chainModule := range expectedChainModules {
+		_, statErr := os.Stat(filepath.Join(path, "vue", "src", "store", "generated", "test", "blog", chainModule))
+		require.False(t, os.IsNotExist(statErr), fmt.Sprintf("the %s vuex store should have be generated", chainModule))
```
Yeah, I will fix this for the rest as well. Oversight from my part. I wrote the tests a bit hurriedly first and forgot about this when checking over it towards the end. Good catch :)

bjaanes commented on starport/pkg/cosmosanalysis/app/app.go:
```diff
@@ -69,3 +70,147 @@ func CheckKeeper(path, keeperName string) error {
+	return basicModules, nil
+}
+
+func extractImports(f ast.Node) map[string]string {
```
Ah yeah. That makes sense. I'll write some extra tests for it at the same time

bjaanes commented on starport/pkg/cosmosanalysis/app/app.go:
```diff
@@ -69,3 +70,147 @@ func CheckKeeper(path, keeperName string) error {
+
+func extractImports(f ast.Node) map[string]string {
+	importMap := make(map[string]string) // name -> import
+	ast.Inspect(f, func(n ast.Node) bool {
```
:facepalm: Yes! Yes we can, much simpler and cleaner

bjaanes commented on starport/pkg/cosmosanalysis/app/app.go:
```diff
@@ -69,3 +70,147 @@ func CheckKeeper(path, keeperName string) error {
+		selectorExprType, ok := callExprType.Fun.(*ast.SelectorExpr)
+		if ok {
+			identExprType, ok := selectorExprType.X.(*ast.Ident)
+			if ok && identExprType.Name == "module" && selectorExprType.Sel.Name == "NewBasicManager" {
```
Oh yeah, good point! I'm adding a lookup from the packages for the named package for "github.com/cosmos/cosmos-sdk/types/module". You think that covers it well enough?

bjaanes commented:
@ilgooz Thanks for the review! I've updated the PR with the changes requested. I also added some tests for other methods in goanalysis since I was already adding some tests there. Hope that is OK :)

ilgooz commented on starport/pkg/cosmosanalysis/app/app.go:
```diff
@@ -69,3 +70,161 @@ func CheckKeeper(path, keeperName string) error {
+
+	var basicModules []string
+	for _, pkg := range pkgs {
+		for p, f := range pkg.Files {
```
There is an error from the integration tests about `findBasicManagerModule`. To fix, maybe instead of checking all files, we can discover the path to `app/app.go` and only operate in this file?

bjaanes replied:
I wasn't able to reproduce the integration test failure locally, but I extended the unit test to find it. The loop now ignore any non app.go file.

bjaanes commented on starport/pkg/cosmosanalysis/app/app.go:
```diff
@@ -69,3 +70,161 @@ func CheckKeeper(path, keeperName string) error {
 }
+
+// FindRegisteredModules looks for all the registered modules in the App
+// It connects the imported go module to the registered module and returns a list of the go modules that are registered
```
Yeah, that reads much better :) Thanks!

bjaanes commented:
@ilgooz I've updated the PR with the requested changes :)

ilgooz reviewed (commented):
Thanks for the updates, added one more comment.

bjaanes commented on starport/pkg/cosmosanalysis/app/app.go:
```diff
@@ -69,3 +73,165 @@ func CheckKeeper(path, keeperName string) error {
+	var basicModules []string
+	for _, pkg := range pkgs {
+		for p, f := range pkg.Files {
+			if !strings.HasSuffix(p, appFileName) {
```
Sounds reasonable! I will do that.
Quick question on the FindAppFilePath, should it take in the path to look for app.go in, or will it do a deep find or assume ./app/* when it searches?

bjaanes replied:
Yeah, that will work. I ended up sending in the chain root since we already have it and just going from "." doesn't work out when scaffolding the chain (and probably not when using --path either). But it seems to be working fine. 

Update coming soon.

bjaanes replied:
@ilgooz I updated the PR now with some changes to account for this. I also added some more tests cosmosanalysis since I was doing some refactoring to be able to reuse the logic already there. Let me know if it makes sense or you want any additional changes to it.

ilgooz replied:
That's great!

ilgooz reviewed (approved):
LGTM! This is a great implementation! Thank you for addressing the change requests and adding extra tests! 🎊 ❤️

ilgooz commented:
We were going to make a hot release and I misconfigured the merge process for this PR. To fix my mistake, I created another PR for your changes and directly merged to develop. 👍  https://github.com/tendermint/starport/pull/2159
